From f8c9df6fdbd11727d560b361898640d1413fe0e2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Patrick=20Sodr=C3=A9?= <psodre@gmail.com>
Date: Tue, 25 Dec 2018 07:39:45 +0000
Subject: [PATCH 2/2] Add byobu-conda command to aid on conda related quirks

  - The byobu environment is activate within every pane.
  - The base environment is stacked on top of byobu.
  - The changes to .bashrc are restricted to byobu sessions only.
---
 configure.ac                   |  3 ++
 usr/bin/Makefile.am            |  2 +-
 usr/bin/byobu-conda.in         | 39 ++++++++++++++++++++++
 usr/bin/byobu-disable-conda.in | 41 +++++++++++++++++++++++
 usr/bin/byobu-enable-conda.in  | 60 ++++++++++++++++++++++++++++++++++
 usr/lib/byobu/include/dirs.in  | 53 ++++++++++++++++++++++++++++++
 6 files changed, 197 insertions(+), 1 deletion(-)
 create mode 100755 usr/bin/byobu-conda.in
 create mode 100755 usr/bin/byobu-disable-conda.in
 create mode 100755 usr/bin/byobu-enable-conda.in

diff --git a/configure.ac b/configure.ac
index 552239d8..ed83055e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -22,11 +22,14 @@ AC_OUTPUT(Makefile \
 	etc/profile.d/Makefile \
 	etc/profile.d/Z97-byobu.sh \
 	usr/bin/byobu \
+	usr/bin/byobu-conda \
 	usr/bin/byobu-config \
 	usr/bin/byobu-ctrl-a \
 	usr/bin/byobu-disable \
+	usr/bin/byobu-disable-conda \
 	usr/bin/byobu-disable-prompt \
 	usr/bin/byobu-enable \
+	usr/bin/byobu-enable-conda \
 	usr/bin/byobu-enable-prompt \
 	usr/bin/byobu-export \
 	usr/bin/byobu-janitor \
diff --git a/usr/bin/Makefile.am b/usr/bin/Makefile.am
index fa33a26a..0d697ea1 100644
--- a/usr/bin/Makefile.am
+++ b/usr/bin/Makefile.am
@@ -1,4 +1,4 @@
-bin_SCRIPTS = byobu  byobu-config  byobu-ctrl-a  byobu-disable  byobu-disable-prompt  byobu-enable  byobu-enable-prompt  byobu-export  byobu-janitor  byobu-keybindings  byobu-launch  byobu-launcher  byobu-launcher-install  byobu-launcher-uninstall  byobu-layout  byobu-prompt  byobu-quiet  byobu-reconnect-sockets  byobu-select-backend  byobu-select-profile  byobu-select-session  byobu-silent  byobu-status  byobu-status-detail  byobu-shell byobu-ugraph byobu-ulevel col1 ctail wifi-status vigpg manifest purge-old-kernels
+bin_SCRIPTS = byobu  byobu-conda  byobu-config  byobu-ctrl-a  byobu-disable  byobu-disable-conda  byobu-disable-prompt  byobu-enable  byobu-enable-conda  byobu-enable-prompt  byobu-export  byobu-janitor  byobu-keybindings  byobu-launch  byobu-launcher  byobu-launcher-install  byobu-launcher-uninstall  byobu-layout  byobu-prompt  byobu-quiet  byobu-reconnect-sockets  byobu-select-backend  byobu-select-profile  byobu-select-session  byobu-silent  byobu-status  byobu-status-detail  byobu-shell byobu-ugraph byobu-ulevel col1 ctail wifi-status vigpg manifest purge-old-kernels
 
 install-exec-hook:
 	cd $(DESTDIR)$(bindir) && rm -f byobu-screen && $(LN_S) byobu byobu-screen
diff --git a/usr/bin/byobu-conda.in b/usr/bin/byobu-conda.in
new file mode 100755
index 00000000..63800b76
--- /dev/null
+++ b/usr/bin/byobu-conda.in
@@ -0,0 +1,39 @@
+#!/bin/sh -e
+#
+#    byobu-conda
+#    Copyright (C) 2013-2014 Dustin Kirkland
+#    Copyright (C) 2018-2019 Patrick Sodré
+#
+#    Authors: Dustin Kirkland <kirkland@byobu.org>
+#             Patrick Sodré <psodre@gmail.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, version 3 of the License.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+PKG="byobu"
+[ -r "$HOME/.byoburc" ] && . "$HOME/.byoburc"
+[ -z "${BYOBU_PREFIX}" ] && export BYOBU_PREFIX="@prefix@" || export BYOBU_PREFIX
+. "${BYOBU_PREFIX}/lib/${PKG}/include/common"
+
+echo
+echo -n "Do you want to enable Byobu's conda integration? [Y/n]: "
+answer=$(head -n1)
+echo
+case "$answer" in
+	"n"|"N")
+		exec byobu-disable-conda
+	;;
+	*)
+		exec byobu-enable-conda
+esac
+
+# vi: syntax=sh ts=4 noexpandtab
diff --git a/usr/bin/byobu-disable-conda.in b/usr/bin/byobu-disable-conda.in
new file mode 100755
index 00000000..e2a96845
--- /dev/null
+++ b/usr/bin/byobu-disable-conda.in
@@ -0,0 +1,41 @@
+#!/bin/sh -e
+#
+#    byobu-disable-conda
+#    Copyright (C) 2013-2014 Dustin Kirkland
+#    Copyright (C) 2018-2019 Patrick Sodré
+#
+#    Authors: Dustin Kirkland <kirkland@byobu.org>
+#             Patrick Sodré <psodre@gmail.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, version 3 of the License.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+PKG="byobu"
+[ -r "$HOME/.byoburc" ] && . "$HOME/.byoburc"
+[ -z "${BYOBU_PREFIX}" ] && export BYOBU_PREFIX="@prefix@" || export BYOBU_PREFIX
+. "${BYOBU_PREFIX}/lib/${PKG}/include/common"
+
+$BYOBU_SED_INLINE -e "/#byobu-conda#$/d" "$HOME/.bashrc"
+[ -w "$HOME/.byoburc" ] && $BYOBU_SED_INLINE -e "/#byobu-conda#$/d" "$HOME/.byoburc"
+
+if [ "$1" != "--no-reload" ]; then
+	if [ -n "$TMUX" ] && [ "$SHELL" = "/bin/bash" ]; then
+		tmux send-keys " . ~/.bashrc" \; send-keys Enter
+	else
+		echo
+		echo "You will need to reload your shell configuration for this to take effect..."
+		echo "  . ~/.bashrc"
+		echo
+	fi
+fi
+
+# vi: syntax=sh ts=4 noexpandtab
diff --git a/usr/bin/byobu-enable-conda.in b/usr/bin/byobu-enable-conda.in
new file mode 100755
index 00000000..e408506b
--- /dev/null
+++ b/usr/bin/byobu-enable-conda.in
@@ -0,0 +1,60 @@
+#!/bin/sh -e
+#
+#    byobu-enable-conda
+#    Copyright (C) 2013-2014 Dustin Kirkland
+#    Copyright (C) 2018-2019 Patrick Sodré
+#
+#    Authors: Dustin Kirkland <kirkland@byobu.org>
+#             Patrick Sodré <psodre@gmail.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, version 3 of the License.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+[ "$SHELL" = "/bin/bash" ] || exit 1
+
+PKG="byobu"
+[ -r "$HOME/.byoburc" ] && . "$HOME/.byoburc"
+[ -z "${BYOBU_PREFIX}" ] && export BYOBU_PREFIX="@prefix@" || export BYOBU_PREFIX
+. "${BYOBU_PREFIX}/lib/${PKG}/include/common"
+
+
+prompt="${BYOBU_CONFIG_DIR}/conda"
+# This gets the conda function into the shell
+echo ". ${BYOBU_CONDA_ROOT}/etc/profile.d/conda.sh #byobu-conda#" > "$prompt"
+# This gets byobu back in the path
+echo "_conda_reactivate >/dev/null 2>&1 || __conda_reactivate >/dev/null 2>&1 #byobu-conda#" >> "$prompt"
+if [ "${BYOBU_PREFIX}" != "${BYOBU_CONDA_ROOT}" ]; then
+	# This gets conda base reactivated with higher precedence than byobu.
+	echo "conda activate ${BYOBU_CONDA_ROOT} #byobu-conda#" >> "$prompt"
+
+	echo
+	echo "Byobu was not installed to the base conda environment."
+	echo "Please increase conda's max shell level stacking to at least 3."
+	echo "  conda config --set max_shlvl 3"
+	echo
+fi
+
+
+$PKG-disable-conda --no-reload "$1" || true
+printf "[ -n \"\$BYOBU_PREFIX\" ] && [ -r $prompt ] && . $prompt   #byobu-conda#\n" >> "$HOME/.bashrc"
+printf "BYOBU_CONDA_ROOT=$BYOBU_CONDA_ROOT    #byobu-conda#\n" >> $HOME/.byoburc
+
+if [ -n "$TMUX" ]; then
+	tmux send-keys " . ~/.bashrc" \; send-keys Enter
+else
+	echo
+	echo "You will need to reload your shell configuration for this to take effect..."
+	echo "  . ~/.bashrc"
+	echo
+fi
+
+# vi: syntax=sh ts=4 noexpandtab
diff --git a/usr/lib/byobu/include/dirs.in b/usr/lib/byobu/include/dirs.in
index 13abf3a3..dabf9f2d 100755
--- a/usr/lib/byobu/include/dirs.in
+++ b/usr/lib/byobu/include/dirs.in
@@ -24,6 +24,59 @@ PKG="byobu"
 [ -r "$HOME/.byoburc" ] && . "$HOME/.byoburc"
 [ -n "$BYOBU_PREFIX" ] || BYOBU_PREFIX="@prefix@"
 
+# https://stackoverflow.com/a/79301903
+is_function() {
+	[ -n "$1" ] && [ "$(command -v -- "$1")" = "$1" ] || return 1
+	[ "$(command -V -- "$1")" != "$(unset -f -- "$1"; command -V -- "$1" 2>/dev/null)" ]
+}
+
+# Some users install byobu using conda/(micro)mamba but the byobu environment is not activated
+if ! [ "$CONDA_PREFIX" -ef "$BYOBU_PREFIX" ]; then
+	# Enable the conda/(micro)mamba command
+	conda_sh="etc/profile.d/conda.sh"
+	if is_function conda || is_function mamba || is_function micromamba; then
+		# conda/(micro)mamba: is a function (already initialized)
+		:
+	elif [ -n "$BYOBU_CONDA_ROOT" ] && [ -r "${BYOBU_CONDA_ROOT}/${conda_sh}" ]; then
+		# conda: user defined location
+		. "${BYOBU_CONDA_ROOT}/${conda_sh}"
+	elif [ -n "$CONDA_EXE" ] && [ -r "$(cd $(dirname "$CONDA_EXE")/..; pwd)/${conda_sh}" ]; then
+		# conda: shell variable defined executable
+		. "$(cd $(dirname "$CONDA_EXE")/..; pwd)/${conda_sh}"
+	elif [ -r "${BYOBU_PREFIX}/${conda_sh}" ]; then
+		# conda: byobu is installed on the base environment
+		. "${BYOBU_PREFIX}/${conda_sh}"
+	elif [ -r "${BYOBU_PREFIX}/../../${conda_sh}" ]; then
+		# conda: byobu is installed on environments from the root
+		. "${BYOBU_PREFIX}/../../${conda_sh}"
+	elif [ -r "/${conda_sh}" ]; then
+		# conda: sysadmin configured location
+		. "/${conda_sh}"
+	elif [ -n "$MAMBA_EXE" ] && [ -x "$MAMBA_EXE" ]; then
+		# (micro)mamba: shell variable defined executable
+		eval "$("$MAMBA_EXE" shell hook -s posix)"
+	elif command -v mamba > /dev/null 2>&1; then
+		# mamba: is a command (not yet initialized)
+		eval "$(mamba shell hook -s posix)"
+	elif command -v micromamba > /dev/null 2>&1; then
+		# micromamba: is a command (not yet initialized)
+		eval "$(micromamba shell hook -s posix)"
+	else
+		echo "WARNING: Please source etc/profile.d/conda.sh first."
+		echo "WARNING: Alternatively, activate the byobu environment and call byobu-enable-conda"
+	fi
+	# Activate the byobu environment using conda/(micro)mamba
+	if is_function conda; then
+		conda activate "$BYOBU_PREFIX"
+	elif is_function mamba; then
+		mamba activate "$BYOBU_PREFIX"
+	elif is_function micromamba; then
+		micromamba activate "$BYOBU_PREFIX"
+	fi
+fi
+BYOBU_CONDA_ROOT=$_CONDA_ROOT
+unset -f is_function
+
 # Create and export the user configuration directory
 # BYOBU_CONFIG_DIR if set, ~/.byobu if it exists, otherwise XDG_CONFIG_HOME/byobu
 if [ -z "$BYOBU_CONFIG_DIR" ]; then
